indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java

```diff
@@ -0,0 +1,377 @@
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * Abstract class for batch tasks like {@link IndexTask}.
+ * Provides some methods ({@link #tryLockWithIntervals} and {@link #tryLockWithSegments}) for easily acquiring task
+ * locks.
+ */
+public abstract class AbstractBatchIndexTask extends AbstractTask
+{
+  @Nullable
+  private Map<Interval, OverwritingRootGenerationPartitions> overwritingRootGenPartitions;
+  @Nullable
+  private Set<DataSegment> allInputSegments;
+
+  @Nullable
+  private Boolean changeSegmentGranularity;
+
+  public static class OverwritingRootGenerationPartitions
+  {
+    private final int startRootPartitionId;
+    private final int endRootPartitionId;
+    private final short maxMinorVersion;
+
+    private OverwritingRootGenerationPartitions(int startRootPartitionId, int endRootPartitionId, short maxMinorVersion)
+    {
+      this.startRootPartitionId = startRootPartitionId;
+      this.endRootPartitionId = endRootPartitionId;
+      this.maxMinorVersion = maxMinorVersion;
+    }
+
+    public int getStartRootPartitionId()
+    {
+      return startRootPartitionId;
+    }
+
+    public int getEndRootPartitionId()
+    {
+      return endRootPartitionId;
+    }
+
+    public short getMinorVersionForNewSegments()
+    {
+      return (short) (maxMinorVersion + 1);
+    }
+  }
+
+  protected AbstractBatchIndexTask(String id, String dataSource, Map<String, Object> context)
+  {
+    super(id, dataSource, context);
+  }
+
+  protected AbstractBatchIndexTask(
+      String id,
+      @Nullable String groupId,
+      @Nullable TaskResource taskResource,
+      String dataSource,
+      @Nullable Map<String, Object> context
+  )
+  {
+    super(id, groupId, taskResource, dataSource, context);
+  }
+
+  public abstract boolean requireLockInputSegments();
+
+  public abstract List<DataSegment> findInputSegments(TaskActionClient taskActionClient, List<Interval> intervals)
+      throws IOException;
+
+  public abstract boolean checkIfChangeSegmentGranularity(List<Interval> intervalOfExistingSegments);
+
+  public abstract boolean isPerfectRollup();
+
+  /**
+   * Returns the segmentGranularity for the given interval. Usually tasks are supposed to return its segmentGranularity
+   * if exists. The compactionTask can return different segmentGranularity depending on its configuration and the input
+   * interval.
+   *
+   * @return segmentGranularity or null if it doesn't support it.
+   */
+  @Nullable
+  public abstract Granularity getSegmentGranularity(Interval interval);
+
+  protected boolean tryLockWithIntervals(TaskActionClient client, List<Interval> intervals, boolean isInitialRequest)
```

Comment:
I'm not sure if this is the most appropriate place for this comment, but should there be some additional configuration that allows/forces continuing to use time chunk locks, so as to be able to roll-back to previous versions? In other words, I think that this segment locking behavior should probably be opt-in; it's a rather strong position to take that by default upgrading will mean that you are unable to roll-back. I think it's probably much safer to retain the existing behavior by default, and then change that default in a future version where rollback will not be problematic.

Reply:
Also agree with @jon-wei and would like to request for more javadocs on the methods in this class

Reply:
> In other words, I think that this segment locking behavior should probably be opt-in;

This sounds nice! Added a task context `forceTimeChunkLock` which is set to true by default. Also added more javadoc to this class.